### PR TITLE
Problem: no stateful precompiled contract for relayer

### PR DESCRIPTION
**Pull request by mmsqe**
Updates: https://github.com/cosmos/relayer/issues/1187

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: integration_tests/ibc_utils.py [hunk 1]
```diff
@@ -16,9 +17,10 @@ class IBCNetwork(NamedTuple):
     chainmain: Chainmain
     hermes: Hermes
     incentivized: bool
+    proc: subprocess.Popen[bytes] | None
 
 
-def prepare_network(tmp_path, file, incentivized=True, start_relay=True):
+def prepare_network(tmp_path, file, incentivized=True, is_relay=True, is_hermes=True):
```
Comment (yihuang) [inline integration_tests/ibc_utils.py:23]: I guess it can't do both, so maybe `relayer = 'hermes | gorelayer'` is better?

**Comment by thomas-nguy** [comment]
I guess this PR is to be able to test the ibc middleware that will convert cosmos msgs to ethermint msgs?

would it be possible to write an ADR for the design? @yihuang

**Comment by yihuang** [comment]
I guess this PR itself is just try go-relayer in integration test, so it'll be more convenient for @mmsqe to further experiment using the golang relayer

**Comment by thomas-nguy** [comment]
yes I understand that this PR is only to integrate with the go-relayer but I just using this opportunity to raise the need for an ADR before getting too deep in the implementation so that we can have a better visibility on what has to be done

**Comment by thomas-nguy** [comment]
Ideally the IBC protocol should be relayer agnostic so using hermes or gorelayer unless we change the semantic for cronos which may raise some concerns (which could be discussed on the ADR)

**Comment by yihuang** [comment]
> Ideally the IBC protocol is relayer agnostic so using hermes or gorelayer implementation should not matter unless we change the semantic for cronos which may raise some concerns (which could be discussed in the ADR)

@mmsqe need to patch the relayer to let it send evm tx instead of cosmos tx.

**Comment by mmsqe** [comment]
> > Ideally the IBC protocol is relayer agnostic so using hermes or gorelayer implementation should not matter unless we change the semantic for cronos which may raise some concerns (which could be discussed in the ADR)
> 
> @mmsqe need to patch the relayer to let it send evm tx instead of cosmos tx.

sorry that I am not so familiar with rust.

**Comment by thomas-nguy** [comment]
```
@mmsqe need to patch the relayer to let it send evm tx instead of cosmos tx.
```

Could it be done without changing the relayer implementation?

something like 

Relayer send cosmos txs -> Cronos middleware catch the cosmos txs and wrap it in an evm txs using module private key 

in a general way, if the concern is to generate the receipt for cosmos txs, I think execution is not needed but we could refactor ethermint (https://github.com/evmos/ethermint/blob/main/x/evm/keeper/msg_server.go#L92) to be able to generate more easily receipts out of an eth txs

**Comment by yihuang** [comment]
> ```
> @mmsqe need to patch the relayer to let it send evm tx instead of cosmos tx.
> ```
> 
> Could it be done without changing the relayer implementation?
> 
> something like
> 
> Relayer send cosmos txs -> Cronos middleware catch the cosmos txs and wrap it in an evm txs using module private key
> 

can you elaborate? by middleware do you mean a rpc service? it can convert the original tx to an evm tx, but can't re-sign the evm tx using the relayer's key, right?

> in a general way, if the concern is to generate the receipt for cosmos txs, I think execution is not needed but we could refactor ethermint (https://github.com/evmos/ethermint/blob/main/x/evm/keeper/msg_server.go#L92) to be able to generate more easily receipts out of an eth txs

there are many things to emulate: tx list in `eth_getBlock`, tx body in `eth_getTransaction`

**Comment by thomas-nguy** [comment]
```
can you elaborate? by middleware do you mean a rpc service? it can convert the original tx to an evm tx, but can't re-sign the evm tx using the relayer's key, right?
```

I mean ibc middleware, or it can even be done in the ante handler if we want to generalise to all cosmos txs

Yes, the evm txs will be signed by an arbitrary key (module key in genesis config) not the relayer key
but I think it is not too important (maybe?) because :
- In case of ibc, what trigger the "change of state" is the relayer, not the original ibc sender and in most of the case, we don't care about the relayer identity (the from field would be irrelevant in most explorer)
- If we need to get the relayer address, we can add an event that will specify it so that it will still be possible to get the oroginal cosmos signer from the explorer

```
there are many things to emulate: tx list in eth_getBlock, tx body in eth_getTransaction, and in the end, we can't emulate the v,r,s signature in tx body.
```

I think we can wrap the cosmos txs in the evm txs like putting the cosmos txs encoded in the call data
 v,r,s signature in tx body will be generated by the module private key
 
 But in some case, we may need to also have the ability to generate "custom" ethereum events which make a bit complex.

**Comment by yihuang** [comment]
> I mean ibc middleware

do you mean broadcast a new tx in the ibc middleware? that is part of the state machine.

**Comment by thomas-nguy** [comment]
`do you mean broadcast a new tx in the ibc middleware? that is part of the state machine.`

One step is to refactor ethermint to make it easier to generate ethereum receipt.

I understand that it is a bit complicated now because its all across many different place. But from my understanding the ethereum receipt information is stored in the cosmos events in many places and reconstructed in the rpc endpoints.

It may be possible to refactor the code to have an utility function to generate receipt. 
(its basically the same as executing ethmsg except that we can ignore the applymsg part)

but maybe there is more complexity (?), I havent look too deep into the code

**Comment by yihuang** [comment]
> `do you mean broadcast a new tx in the ibc middleware? that is part of the state machine.`
> 
> One step is to refactor ethermint to make it easier to generate ethereum receipt.
> 
> I understand that it is a bit complicated now because its all across many different place. But from my understanding the ethereum receipt information is stored in the cosmos events in many places and reconstructed in the rpc endpoints.
> 
> It may be possible to refactor the code to have an utility function to generate receipt. (its basically the same as executing ethmsg except that we can ignore the applymsg part)
> 
> but maybe there is more complexity (?), I havent look too deep into the code

even if we emulated the events, there are still some difficulties, for example `getBlock` filter evm txs by msg type and result code, now it need to look into the event list of every tx, and in the end we still can't emulate the signature of the evm tx.

**Comment by thomas-nguy** [comment]
another possible flow is 

cosmos tx -> middleware convert to eth tx signed with module private key -> execute the eth txs as a specific precompile (we may be able to bypass signature verification check and set direction From to original cosmos sender)

at the end we can get the correct tx result code and gas used. 

but not sure if its possible to change to bypass the signature verification without any risks?

**Comment by yihuang** [comment]
> another possible flow is
> 
> cosmos tx -> middleware convert to eth tx signed with module private key -> execute the eth tx as a specific precompile (we may be able to bypass signature verification check and set direction From to original cosmos sender)
> 
> at the end we can get the correct tx result code and gas used in the receipt.
> 
> but not sure if its possible to change to bypass the signature verification without any risks?

I guess ibc incentivization feature rely on correct signer of the tx.

**Comment by mmsqe** [comment]
@calvinaco it sign normal evm tx, see if we need support dynamic fee tx

File: x/cronos/keeper/precompiles/relayer.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	BaseContract
+
+	cdc       codec.Codec
+	ibcKeeper *ibckeeper.Keeper
+}
+
+func NewRelayerContract(ibcKeeper *ibckeeper.Keeper, cdc codec.Codec) precompiles.StatefulPrecompiledContract {
+	return &RelayerContract{
+		BaseContract: NewBaseContract(
+			common.BytesToAddress([]byte{101}),
```
Comment (yihuang) [inline x/cronos/keeper/precompiles/relayer.go:32]: define a global constant variable for `101`?

File: x/cronos/keeper/precompiles/relayer.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		),
+		ibcKeeper: ibcKeeper,
+		cdc:       cdc,
+	}
+}
+
+func (bc *RelayerContract) Address() common.Address {
+	return common.BytesToAddress([]byte{101})
```
Comment (yihuang) [inline x/cronos/keeper/precompiles/relayer.go:40]: https://github.com/crypto-org-chain/cronos/pull/1014/files#r1320933333

File: x/cronos/keeper/precompiles/relayer.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// RequiredGas calculates the contract gas use
+func (bc *RelayerContract) RequiredGas(input []byte) uint64 {
+	// TODO estimate required gas
+	return 0
```
Comment (yihuang) [inline x/cronos/keeper/precompiles/relayer.go:46]: use a constant number for now?

File: nix/sources.json [hunk 5]
```diff
@@ -84,6 +84,18 @@
         "url": "https://github.com/informalsystems/ibc-rs/archive/daad02843a091bbdb3dd608e5f4ce790895c8845.tar.gz",
         "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
     },
+    "relayer": {
+        "branch": "master",
+        "description": "An IBC relayer for ibc-go",
+        "homepage": "https://github.com/mmsqe/relayer",
```
Comment (yihuang) [inline nix/sources.json:90]: replace with the fork after merged: https://github.com/crypto-org-chain/relayer/

File: x/cronos/keeper/precompiles/relayer.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	prefixChannelOpenConfirm
+	prefixChannelCloseInit
+	prefixChannelCloseConfirm
+	prefixRecvPacket
+	prefixAcknowledgement
+	prefixTimeout
+	prefixTimeoutOnClose
+)
+
+// exec is a generic funciton that executes the given action in statedb, and marshal/unmarshal the input/output
```
Comment (mmsqe) [inline x/cronos/keeper/precompiles/relayer.go:80]:
```suggestion
// exec is a generic function that executes the given action in statedb, and marshal/unmarshal the input/output
```

File: x/cronos/keeper/precompiles/relayer.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	conntypes "github.com/cosmos/ibc-go/v7/modules/core/03-connection/types"
+	chantypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	ibckeeper "github.com/cosmos/ibc-go/v7/modules/core/keeper"
+)
+
+// TODO adjust the gas cost
+const RelayerContractRequiredGas = 1000000
```
Comment (yihuang) [inline x/cronos/keeper/precompiles/relayer.go:22]:
```suggestion
const RelayerContractRequiredGas = 10000
```

**Review by yihuang — APPROVED** [review]

**Pull request merged**